### PR TITLE
Amend backorder after cancellations

### LINKED ISSUE #12876: [DFC Orders] Cancelled Order does not reduce DFC Order quantities immediately
### Testing issue for #12856 
When an order is cancelled that contains DFC products, stock is returned but the quantity ordered from the Producer is not adjusted down.

## Expected Behaviour
When an order is cancelled (& items restocked), if that variant now has the same or more (>=) stock than the wholesale case, the DFC Order should be reduced by 1 case.

## Actual Behaviour
Items were restocked, no adjustment was made to the DFC Order, so too many wholesale cases were ordered.

## Steps to Reproduce

1. Create hub
2. Link OIDC account (ensure authorised to DFC Shopify Store, including setting up Shopify account to match OIDC User Id/email)
3. Import products
4. Create Order Cycle including DFC products
5. Place an order for DFC Products
6. Confirm order has created draft order on Shopify
7. Cancel Order & restock items
8. Confirm Shopify draft order has not been adjusted.

### Cancelled order: 
image
### Over-stocked product:
image
### Shopify draft Order:
image

## Severity
bug-s2: a non-critical feature is broken, no workaround

## Your Environment

* Version used: uk-staging
* Browser name and version: Firefox 130.0.1 (64 bit)
* Operating System and version (desktop or mobile): Ubuntu 22.04, desktop

## Possible Fix
Need to adjust (update) DFC Order on OFN Order cancellation (where stock >= case quantity).

### PR DESCRIPTION
**Pull request by mkllnk**
**:information_source:  Funded Feature. Please track ALL ASSOCIATED WORK under the associated tracking code: #11678 DFC Orders**

#### What? Why?

- Closes #12876 

The new job class blends code from the BackorderJob and the
CompleteBackorderJob for the specific case of adjusting quantities after
an order has been cancelled.

I would like to write a more general class which can be used for any
order amendmends but this was the quickest solution to cater for
currently running pilots.

#### What should we test?

- For @RaggedStaff:
- Checkout -> backorder is created.
- Cancel order -> backorder should be adjusted. If there was only one order then this results in an empty order, I think. I hope that doesn't cause trouble down the line. I could add a check and set it to cancelled?

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [x] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by RaggedStaff** [comment]
Ok, I've tested this & it seems to work (Shopify draft orders are correctly amended when an OFN order is cancelled)! 🎉

**Review by rioug — APPROVED** [review]
A bit hard to follow all the transformations, but it makes sense in the end. Nice one !

File: spec/jobs/amend_backorder_job_spec.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      # Same as before:
+      expect(beans.on_hand).to eq 9
+      expect(chia_seed.on_hand).to eq 7
+      expect(backorder.lines[0].quantity).to eq 1 # beans
+      expect(backorder.lines[1].quantity).to eq 5 # chia
+
+      # We cancel the only order and that should reduce the order lines to 0.
+      expect { order.cancel! }
+        .to change { beans.reload.on_hand }.from(9).to(15)
+        .and change { chia_seed.reload.on_hand }.from(7).to(12)
```
Comment (dacook) [inline spec/jobs/amend_backorder_job_spec.rb:92]: If the retail order and the backorder are both cancelled, why would the on_hand be changed?

**Review by dacook — APPROVED** [review]
I have to admit I don't completely follow it, but I can see that it's been clearly specced and confirmed with manual testing already, and it needs to move forward.

Also I think it will be important to refactor in the not-too-distant future to ensure integrity (so we don't have duplicated code get out of sync) and make it easier for developers to use and understand it.

File: spec/jobs/amend_backorder_job_spec.rb (hunk 1 above)
Comment (mkllnk) [reply spec/jobs/amend_backorder_job_spec.rb:92]: This block is cancelling the OFN order. So when you cancel the order the local OFN stock is returned. The backorder hasn't been touch yet. That's done in the next block.

When the backorder is updated, there are two different methods. If the product is on demand then the local stock may be reduced again because we are cancelling part of the backorder at the same time. But when it's stock controlled, we don't touch the OFN stock. The backorder is only affected by the line item quantities while the stock is synced with the remote catalog.
